### PR TITLE
refactor ibc40 middleware 

### PR DESCRIPTION
**Pull request by ItsFunny**
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/types/decoder.go [hunk 1]
```diff
@@ -61,6 +61,11 @@ func TxDecoder(cdc codec.CdcAbstraction) sdk.TxDecoder {
 			if tx, err = f(cdc, txBytes, height); err == nil {
 				tx.SetRaw(txBytes)
 				tx.SetTxHash(types.Tx(txBytes).Hash(height))
+				if seisitive, ok := tx.(sdk.HeightSensitive); ok {
```
Comment (zjg555543) [inline x/evm/types/decoder.go]: 1、seisitive -> sensitive?
2、All tx will check, whether it’s efficient？
3、Add some ut for this code？

**Review by zjg555543 — APPROVED** [review]

File: libs/cosmos-sdk/types/coin_ibc.go [hunk 2]
```diff
@@ -308,3 +313,93 @@ func (coins CoinAdapters) Copy() CoinAdapters {
 
 	return copyCoins
 }
+
```
Comment (BananaLF) [inline libs/cosmos-sdk/types/coin_ibc.go:314]: https://github.com/okex/exchain/blob/dev/libs/cosmos-sdk/types/coin_adapter.go exchain has the same func suggest use it

File: libs/cosmos-sdk/x/bank/internal/keeper/keeper_adapter.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package keeper
```
Comment (BananaLF) [inline libs/cosmos-sdk/x/bank/internal/keeper/keeper_adapter.go]: exchain old code have the same keeper_adapter https://github.com/okex/exchain/blob/dev/libs/cosmos-sdk/x/bank/internal/keeperadapter/bankkeeper_adapter.go suggest merge code

File: libs/ibc-go/modules/apps/27-interchain-accounts/types/amino_account.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	return sdkerrors.Wrap(ErrUnsupported, "cannot set sequence number for interchain account")
+}
+
+// Validate implements basic validation of the AminoInterchainAccount
+func (ia AminoInterchainAccount) Validate() error {
+	if strings.TrimSpace(ia.AccountOwner) == "" {
+		return sdkerrors.Wrap(ErrInvalidAccountAddress, "AccountOwner cannot be empty")
+	}
+
+	return ia.BaseAccount.Validate()
```
Comment (BananaLF) [inline libs/ibc-go/modules/apps/27-interchain-accounts/types/amino_account.go:62]: AminoInterchainAccount must be check PubKey == nil and sequence == 0。And this line wile checkout pubkey.address == address. it's maybe a bug

File: libs/cosmos-sdk/types/coin_ibc.go (hunk 2 above)
Comment (ItsFunny) [reply libs/cosmos-sdk/types/coin_ibc.go:314]: done

File: libs/cosmos-sdk/x/bank/internal/keeper/keeper_adapter.go (hunk 3 above)
Comment (ItsFunny) [reply libs/cosmos-sdk/x/bank/internal/keeper/keeper_adapter.go]: done

File: libs/ibc-go/modules/apps/27-interchain-accounts/types/amino_account.go (hunk 4 above)
Comment (ItsFunny) [reply libs/ibc-go/modules/apps/27-interchain-accounts/types/amino_account.go:62]: done

File: app/app.go [hunk 5]
```diff
@@ -502,11 +573,31 @@ func NewOKExChainApp(
...
+	transferStack := ibcporttypes.NewFallThroughMiddleware(left,
+		ibccommon.DefaultFactory(tmtypes.HigherThanVenus4, ibc.IBCV4, right),
+		ibccommon.DefaultFactory(tmtypes.HigherThanVenus1, ibc.IBCV2, middle))
+
 	// Create static IBC router, add transfer route, then set and seal it
 	ibcRouter := ibcporttypes.NewRouter()
-	ibcRouter.AddRoute(ibctransfertypes.ModuleName, transferModule)
+
+	var icaControllerStack ibcporttypes.IBCModule
+	icaMauthIBCModule := icamauth.NewIBCModule(app.ICAMauthKeeper)
```
Comment (BananaLF) [inline app/app.go:586]: why not use ' icaMauthIBCModule := icamauth.NewIBCModule(app.ICAMauthKeeper)
	icaControllerStack := icacontroller.NewIBCMiddleware(icaMauthIBCModule, app.ICAControllerKeeper) ' to replace this codes

File: libs/cosmos-sdk/x/auth/client/utils/tx.go [hunk 6]
```diff
@@ -588,3 +588,19 @@ func isTxSigner(user sdk.AccAddress, signers []sdk.AccAddress) bool {
 
 	return false
 }
+
+func CliConvertCoinToCoinAdapters(coins sdk.Coins) sdk.CoinAdapters {
```
Comment (BananaLF) [inline libs/cosmos-sdk/x/auth/client/utils/tx.go:592]: suggest use CoinsToCoinAdapters

File: libs/ibc-go/modules/core/common/selector.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+///////
+
+type SelectorStrategy struct {
+	internal  interface{}
+	selectors Selectors
+	seal      bool
+}
+
+func (f *SelectorStrategy) RegisterSelectors(factories ...SelectorFactory) {
```
Comment (BananaLF) [inline libs/ibc-go/modules/core/common/selector.go:81]: if f.selectors is not nil, it's will be cover old f.selectors.

File: libs/ibc-go/modules/core/common/selector.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+func NewCommonHeightSelector(higherThan func(h int64) bool, version SelectVersion, internal interface{}) *CommonHeightSelector {
+	return &CommonHeightSelector{higherThan: higherThan, version: version, internal: internal}
+}
+
+func (c *CommonHeightSelector) Version() SelectVersion {
+	return c.version
+}
+
+func (c *CommonHeightSelector) Select(ctx sdk.Context) (interface{}, bool) {
```
Comment (BananaLF) [inline libs/ibc-go/modules/core/common/selector.go:66]: inteface{} may be easy to understand. use type alias is better。use keeper inteface much better

File: libs/ibc-go/modules/apps/29-fee/ibc_middleware.go [hunk 9]
```diff
@@ -61,18 +61,18 @@ func (im IBCMiddleware) OnChanOpenInit(
...
 	}
 
 	versionMetadata.AppVersion = appVersion
 	versionBytes, err := types.ModuleCdc.MarshalJSON(&versionMetadata)
 	if err != nil {
-		return version, err
+		return "", err
 	}
 
 	im.keeper.SetFeeEnabled(ctx, portID, channelID)
```
Comment (BananaLF) [inline libs/ibc-go/modules/apps/29-fee/ibc_middleware.go:78]: This line is enable ics29. But i only wacth ics29 enable during chanel creating. weather we need to enable old channel.

File: app/app.go [hunk 10]
```diff
@@ -502,11 +573,31 @@ func NewOKExChainApp(
 	app.TransferKeeper = *app.TransferKeeper.SetHooks(erc20.NewIBCTransferHooks(app.Erc20Keeper))
 	transferModule := ibctransfer.NewAppModule(app.TransferKeeper, codecProxy)
 
+	left := common.NewDisaleProxyMiddleware()
+	middle := ibctransfer.NewIBCModule(app.TransferKeeper, transferModule)
+	right := ibcfee.NewIBCMiddleware(middle, app.IBCFeeKeeper)
+	transferStack := ibcporttypes.NewFallThroughMiddleware(left,
```
Comment (BananaLF) [inline app/app.go]: transferStack in the gaia code is another ibc_appmodule,why are you change the module. And I saw you code logic is different gaia code

**Review by BananaLF — APPROVED** [review]

**Review by lyh169 — APPROVED** [review]

**Pull request merged**